### PR TITLE
Vpnaas: Create IPSec Policy

### PR DESCRIPTION
**Pull request by simonre**
For #723 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_db.py#L396

https://github.com/openstack/neutron-vpnaas/blob/058469e1b99b647537a5228c6a384d93df5484df/neutron_vpnaas/db/vpn/vpn_db.py#L377

API:
https://developer.openstack.org/api-ref/network/v2/#create-ipsec-policy

### CODE REVIEW

**Comment by simonre** [comment]
@jtopjian this is ready for review

File: openstack/networking/v2/extensions/vpnaas/ipsecpolicies/results.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+type Lifetime struct {
+	// LifetimeUnits is the unit for the lifetime
+	// Default is seconds
+	LifetimeUnits string `json:"units"`
+
+	// LifetimeValue is the lifetime
+	// Default is 3600
+	LifetimeValue int `json:"value"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/ipsecpolicies/results.go]: Let's rename this to `Value`.

File: openstack/networking/v2/extensions/vpnaas/ipsecpolicies/testing/requests_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	th.AssertEquals(t, "b4eedccc6fb74fa8a7ad6b08382b852b", actual.TenantID)
+	th.AssertEquals(t, "ipsecpolicy1", actual.Name)
+	th.AssertEquals(t, "esp", actual.TransformProtocol)
+	th.AssertEquals(t, "sha1", actual.AuthAlgorithm)
+	th.AssertEquals(t, "tunnel", actual.EncapsulationMode)
+	th.AssertEquals(t, "aes-128", actual.EncryptionAlgorithm)
+	th.AssertEquals(t, "group5", actual.PFS)
+	th.AssertEquals(t, "", actual.Description)
+	th.AssertEquals(t, "seconds", actual.Lifetime.LifetimeUnits)
+	th.AssertEquals(t, 7200, actual.Lifetime.LifetimeValue)
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/ipsecpolicies/testing/requests_test.go]: Similar to our other discussion, we'll want to compare full structs. I've tested this out locally and it works, so let me know if you run into problems.

File: openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	//Lifetime is the lifetime of the security association
+	Lifetime *LifetimeCreateOpts `json:"lifetime,omitempty"`
+}
+
+// The lifetime consists of a unit and integer value
+// You can omit either the unit or value portion of the lifetime
+type LifetimeCreateOpts struct {
+	// LifetimeUnits is the units for the lifetime of the security association
+	// Default unit is seconds
+	LifetimeUnits string `json:"units,omitempty"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go]: Let's change this to `Units`.

File: openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+	// The lifetime value.
+	// Must be a positive integer.
+	// Default value is 3600.
+	LifetimeValue int `json:"value,omitempty"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go]: Let's change this to `Value`.

File: openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+type PFS string
+type Unit string
+
+const (
+	TransformProtocolESP       TransformProtocol   = "esp"
+	TransformProtocolAH        TransformProtocol   = "ah"
+	TransformProtocolAHESP     TransformProtocol   = "ah-esp"
+	AuthAlgorithmSHA1          AuthAlgorithm       = "sha1"
+	AuthAlgorithmSHA256        AuthAlgorithm       = "sha256"
+	AuthAlgorithmHA384         AuthAlgorithm       = "sha384"
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/vpnaas/ipsecpolicies/requests.go]: AuthAlgorithmSHA384

**Comment by simonre** [comment]
@jtopjian While writing the List function I noticed that this was missing an 'id' field in the resulting struct so I added it as well.

**Comment by jtopjian** [comment]
@simonre Nice catch - thanks.

It looks like this might need another rebase. The `Vpnaas: Create VPN service (#760)` commit should be omitted because if I squash these commits, including that one commit might make the git history look as though _this_ commit added the Services resource.

**Comment by simonre** [comment]
@jtopjian Is this better?

**Comment by jtopjian** [comment]
@simonre This looks good, but unfortunately the merge of the VPN service delete PR caused a conflict with the `acceptance/openstack/networking/v2/extensions/vpnaas/vpnaas.go` file. Another rebase will resolve this.

Note that when you do the rebase, git will now complain that there's a merge issue. This is expected. You'll want to edit the `vpnaas.go` file and resolve the conflict, then continue with the rebase (`git rebase --continue`).

Let me know if you need help.

Once the conflict is resolved, this is good to go :)

**Comment by jtopjian** [comment]
@simonre This looks good to me. Very nice work on this - especially with the types.

Just a heads up that the `TestPolicyCRUD` acceptance test should probably be renamed to `TestIPSecPolicyCRUD` or else it'll clash with the upcoming `TestPolicyCRUD` in the IKE PRs.

**Pull request merged**